PhysicsTools/MXNet/src/MXNetCppPredictor.cc

```diff
@@ -0,0 +1,146 @@
+/*
+ * MXNetCppPredictor.cc
+ *
+ *  Created on: Jul 19, 2018
+ *      Author: hqu
+ */
+
+#include <cassert>
+#include "FWCore/Utilities/interface/Exception.h"
+
+#include "PhysicsTools/MXNet/interface/MXNetCppPredictor.h"
+
+namespace mxnet {
+
+namespace cpp {
+
+Block::Block() {
+}
+
+Block::Block(const std::string& symbol_file, const std::string& param_file) {
+  // load the symbol
+  sym_ = Symbol::Load(symbol_file);
+  // load the parameters
+  load_parameters(param_file);
+}
+
+Block::~Block() {
+}
+
+void Block::load_parameters(const std::string& param_file) {
+  std::map<std::string, NDArray> paramters;
+  NDArray::Load(param_file, nullptr, &paramters);
+  for (const auto &k : paramters) {
+    if (k.first.substr(0, 4) == "aux:") {
+      auto name = k.first.substr(4, k.first.size() - 4);
+      aux_map_[name] = k.second;
+    }
+    if (k.first.substr(0, 4) == "arg:") {
+      auto name = k.first.substr(4, k.first.size() - 4);
+      arg_map_[name] = k.second;
+    }
+  }
+}
+
+std::mutex MXNetCppPredictor::mutex_;
+const Context MXNetCppPredictor::context_ = Context(DeviceType::kCPU, 0);
+
+MXNetCppPredictor::MXNetCppPredictor() {
+}
+
+MXNetCppPredictor::MXNetCppPredictor(const Block& block) : sym_(block.symbol()), arg_map_(block.arg_map()), aux_map_(block.aux_map()) {
+}
+
+MXNetCppPredictor::MXNetCppPredictor(const Block &block, const std::string &output_node) : sym_(block.symbol(output_node)), arg_map_(block.arg_map()), aux_map_(block.aux_map()) {
+}
+
+MXNetCppPredictor::~MXNetCppPredictor() {
+}
+
+void MXNetCppPredictor::set_input_shapes(const std::vector<std::string>& input_names, const std::vector<std::vector<mx_uint> >& input_shapes) {
+  assert(input_names.size() == input_shapes.size());
+  input_names_ = input_names;
+  // init the input NDArrays and add them to the arg_map
+  for (unsigned i=0; i<input_names_.size(); ++i){
+    const auto& name = input_names_[i];
+    NDArray nd(input_shapes[i], context_, false);
+    arg_map_[name] = nd;
```

Comment:
the static analyzer points to use-after-free error around here https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-23768/29866/llvm-analysis/report-3abc75.html#Path6